### DISCUSSION TITLE
Warning 'An update inside a test was not wrapped in act(...)' while testing with React Testing Library

### DISCUSSION THREAD

**Discussion by antoine-coulebaud** [react-hook-form/react-hook-form#4232] [Q&A]
Hi

I wanted to test some of my react-hook-form components with react-testing-libray.

While some of my tests seems to pass as expected, the warning `An update inside a test was not wrapped in act(...)` pops and I can't figure why. I tried to reduce the code to the minimum in ordre to reproduce, and rewrite it in a sandbox (https://codesandbox.io/s/adoring-lamarr-xxxnj) but infortunately an other error occurred and I can't retrieve the exact same warning as in my own repo.

Can you tell me :
1. How to fix the sandbox ?
2. What's wrong with my code / test and what can trigger this warning ?

Thanks

**Comment by bluebill1049** [comment]
hey @antoine-coulebaud this is mostly due to async validation. you will need to wrap with `async act`, you can take a look at our repo for tests examples.

**Reply by antoine-coulebaud** [reply]
Thanks for the reply, but I do not find any example of component testing on your repository

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.test.tsx#L74

**Reply by vallode** [reply]
@bluebill1049 does this imply we need to wrap `render` in another async act? For me that exact reproduction throws an error on any instance of `register()` being called, with no inputs being fired yet.

Wrapping render just causes a separate warning about overlapping act calls sadly.

**Reply by IanVS** [reply]
Note: the link above is dead.  I'm getting this warning when using a `Controller` similar to what is shown in https://codesandbox.io/s/react-hook-form-parse-and-format-textarea-zbgog?file=/src/index.tsx:1024-1037.  The warning happens when I try to use `userEvent.type()` into my input.

It also occurs when I use the approach from https://github.com/react-hook-form/react-hook-form/blob/master/src/__tests__/controller.test.tsx#L204 like:

```
await act(async () => {
  fireEvent.input(screen.getByRole('textbox'), {
    target: { value: '' },
  });
});
```

**Reply by IanVS** [reply]
OK, I think I've got it working using something like this:

```
const myInput = getByRole('textbox');
userEvent.type(myInput, 'value');
await waitFor(() => expect(myInput).toHaveValue('value'));
```

I had an issue in one case where my expect was satisfied immediately (too early), so make sure that you await something that actually changes!

**Reply by lguardado** [reply]
The link is broken btw, this is the working one today:
https://github.com/react-hook-form/react-hook-form/blob/master/src/__tests__/useForm.test.tsx

**Reply by kathyhuang666** [reply]
@IanVS  Did you find a solution?

**Reply by IanVS** [reply]
The key seems to be using `waitFor` liberally, and being sure that each change is waited for, before making other changes.

**Reply by kathyhuang666** [reply]
Thanks a lot for the replying!

**Reply by ratnajeet-wenergy** [reply]
This worked perfectly for me. Even after I wrapped the previous statement in `act`, it was still throwing the same error. After adding `waitFor` in the next line, it worked perfectly.

**Comment by mAAdhaTTah** [comment]
I'm dealing with this as well. My understanding is the `onChange` mode performs an initial async validation right after render which triggers the warning. Changing modes would ostensibly fix it, but I don't have a fix (yet) for the `onChange` mode.

**Reply by bondarenko09** [reply]
hey @mAAdhaTTah any updates to fix that?

**Reply by mAAdhaTTah** [reply]
No, I just changed modes.

**Reply by brandonlenz** [reply]
I have the same issues with tests that trigger react hook form's onChange handler. If I find a solution for that, I'll update.

**Comment by inPhoenix** [comment]
My issue is when I wrap the fireEvent with the act, it solves the warning, but then I get the following Eslint error:
ESLint: Avoid wrapping Testing Library util calls in `act`(testing-library/no-unnecessary-act)

My code:

```javascript
await act(async () => {
      fireEvent.change(screen.getByLabelText("Email"), {
        target: { value: "validEmail@yopmail.com" },
      });
```

My test verifies if the error is not in the document, and the waitFor solution does not work for me.

**Reply by Belco90** [reply]
Hey @inPhoenix! I'm the author of `eslint-plugin-testing-library`, and I implemented the `no-unnecessary-act` rule.

I found myself in the same situation as you when testing a form built with `react-hook-form`. When waiting for something after triggering an event with Testing Library, but still seeing the act warning, I'm afraid the `act()` is necessary and `no-unnecessary-act` rule has to be suppressed.

I still don't understand what happens inside `react-hook-form` to cause this weird behavior.

**Reply by brandonlenz** [reply]
@Belco90 Closing in on a year later, did you have any success? I'm running into these when calling userEvent methods that trigger the RHF onChange handler, as well as form submission.

So far I have to write

```TSX
await act(() => user.click(submitButton))
```

or

```TSX
await act(() => user.click(yesAnswer))
```

instead of the correct

```TSX
await user.click(submitButton)
```

or

```TSX
await user.click(yesAnswer)
```

respectively.

Hoping someone has a better answer. I'm working on conversion from Formik to RHF. RHF is better in a ton of ways, but this is a huge annoyance when you have lots of tests already.

**Reply by brandonlenz** [reply]
The submitButton one surprises me, since I have an `await waitFor` just afterwards specifically testing that the error validation shows up. But that's what's happening 🤷🏻

**Reply by craigmcc** [reply]
What has worked for me was eliminating the act() wrapper completely, so using things like:

```tsx
        await client.click(save);
```

instead of

```tsx
        await act(async () => {
            await client.click(save);
        });
```
made all those warnings go away.  Maybe the double await was part of the issue?

**Reply by brandonlenz** [reply]
@craigmcc I get warning unless I use act. I started off coding user events the correct way (see [previous answer](https://github.com/orgs/react-hook-form/discussions/4232#discussioncomment-5785875)), but the redundant act resolved the warnings.

I'd love a solution that doesn't involve going against UserEvent best practices, though.

**Reply by pedoch** [reply]
@inPhoenix thank you for this! It worked for me.

**Comment by pengpengzx** [comment]
```js
    await act(async () => {
      render(<MyComponent/>, {
        wrapper: Wrapper,
      });
    });
```
Add `await act (async () => {})`  works for me

**Reply by Belco90** [reply]
That's an anti pattern. You should never wrap the `render` or actions triggered with Testing Library within `act`.

This is why is so strange having to do so when working with this form library. This sort of workarounds shouldn't be needed at all.

You can read more here: https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning

**Reply by pengpengzx** [reply]
Thanks, learned a lot!

**Reply by adroste** [reply]
This is the official recommended way when doing simple snapshot tests, see: https://react-hook-form.com/advanced-usage#TestingForm

**Reply by Belco90** [reply]
I've updated the `react-hook-form` advice about testing forms in [this PR](https://github.com/react-hook-form/documentation/pull/882), should be fine now!

**Reply by bluebill1049** [reply]
thanks for correcting that @Belco90 please refer to the updated doc @adroste (sorry about that incorrect information)

**Reply by liorpevzner** [reply]
Im sorry but why is this an anti-pattern?
If my component is complex and has many api calls on init(which I'm mocking with w/e library I'm using) why wrapping my component with act is an anti-pattern? I want all my async functions to resolve and my DOM to be stable before I start asserts.
Ive read what you sent by kentcdodds (great read, thanks very informative).

**Comment by vcapretz** [comment]
@bluebill1049 @Belco90 hi there folks, thank you for the docs on testing, they really helped 🙏 

I'm still seeing the `...was not wrapped in act...` warnings when trying to destruct values from `formState`.

I have a reproducible example in this [CodeSandbox link](https://codesandbox.io/s/reverent-blackburn-15wfw9?file=/src/App.tsx)
I copied the exact example from the docs page and the only change was

```
const {
    register,
    handleSubmit,
    formState,
} = ...
```
  to 

```
const {
    register,
    handleSubmit,
    formState: { isValid }
  }
```

is there any recommended workaround for this scenario?

**Reply by Belco90** [reply]
[As explained in the testing section of the docs](https://react-hook-form.com/advanced-usage/#TestingForm), you need to wait for the first element of your form/UI to appear. This is due to React Hook Form async nature. By using a `find*` query for your test, the act warning is gone:

```js
it("should have a submit button", async () => {
  render(<App />);

  expect(await screen.findByText("SUBMIT")).toBeInTheDocument();
});
```

**Reply by VictorGlindasPaf** [reply]
This doesn't always work, as the element appears before the validation happens.
So if I test that text in the form appears, validation  still happens and throws "act" errors afterwards.

**Reply by josafatclicktripz** [reply]
Same here, still happens even by using the solution explained in testing section. Only happens when you pass `defaultValues` 😢 Shared CodeSandbox from @vcapretz has the most basic repro steps for this issue.

**Reply by thiago-silva-ac** [reply]
Same here, waiting for element to appear doesn't fix the problem when using "isValid". There is some way of waiting for this validation?

**Reply by neilsot** [reply]
For validation related "act" errors that occur after your last meaningful test assertion what about awaiting unmount?

```jsx
it("should have two text inputs", async () => {
  const { unmount } = render(<Form />);
  
  expect(await screen.findAllByRole('textbox')).toHaveLength(2);
  await unmount();
}
```

Internally it looks like [react-testing-library is wrapping unmount in act()](https://github.com/testing-library/react-testing-library/blob/main/src/pure.js#L154).

I like that this approach prevents me from having to call `act()` manually, or add extra test assertions just to wait for react-hook-form validation. Something I don't like about this approach is that the `await unmount()` might mask other act errors from async effects unrelated to react-hook-form that should have been handled (or mocked).

**Reply by LucasHimelfarb** [reply]
@Belco90 's solution is correct. You should wait for the FIRST element of your Form, in react native it means the first Controller renderer I guess...

**Comment by jackson5cc** [comment]
Having the same issue when using React Hook Form with a Select component in Radix UI. 

If my form has either a Select or a TextInput , everything works like a charm. I'm using the latest user-event library and awaiting all user actions. But if I have both a TextInput and a Select, I get the famous warning. 

Here are my two workarounds: 

- Workaround 1: Select an option from the Select component _before_ typing into the input field. 
- Workaround 2: Call user.tab() after typing into the field. 

I cannot understand why these workarounds work. Can anyone explain it to me?

**Reply by ravicious** [reply]
Omg, thank you. I don't use react-hook-form, but I use HTML validation on an input. I had to `user.tab()` after `user.type()`, otherwise I'd get errors about not wrapping updates after the focus on the page gets changed due to other interactions.

**Comment by the-spyke** [comment]
Is there a way to understand when RHF has completed its validation? Basically, I have a hook that calls `useForm` and a bunch of effects. Adding `await act(async () => {})` helps, but I would prefer to explicitly wait on something. The web site only mentions waiting on a component to appear, but I'm writing test just for the hook.
